spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,499 @@
+    load_train_labels=True, include_dev=False
+)
+
+df_train.head()
+
+
+# %% [markdown]
+# ## 1. Write slicing functions
+#
+# We leverage *slicing functions* (SFs) — an abstraction that shares syntax with *labeling functions*, which you should already be familiar with.
+# (If not, please see the [intro tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/spam/01_spam_tutorial.ipynb).)
+# A key difference: whereas labeling functions output labels, slicing functions output binary _masks_ indicating whether an example is in the slice or not.
+
+# %% [markdown]
+# In the following cells, we use the [`@slicing_function()`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/slicing/snorkel.slicing.slicing_function.html#snorkel.slicing.slicing_function) decorator to initialize an SF that identifies shortened links the spam dataset.
+# These links could redirect us to potentially dangerous websites, and we don't want our users to click them!
+# To select the subset of shortened links in our dataset, we write a regex that checks for the commonly-used `.ly` extension.
+#
+# You'll notice that the slicing function is noisily defined — it doesn't represent the ground truth for all short links.
+# Instead, SFs are often heuristics to quickly measure performance over important subsets of the data.
+
+# %%
+import re
+from snorkel.slicing import slicing_function
+
+
+@slicing_function()
+def short_link(x):
+    """Returns whether text matches common pattern for shortened ".ly" links."""
+    return bool(re.search(r"\w+\.ly", x.text))
+
+
+sfs = [short_link]
+
+# %% [markdown]
+# For our $n$ examples and $k$ slices in each split, we apply the SF to our data to create an $n \times k$ matrix. (So far, $k=1$).
+
+# %%
+from snorkel.slicing import PandasSFApplier
+
+applier = PandasSFApplier(sfs)
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+S_test = applier.apply(df_test)
+
+# %% [markdown]
+# ### Visualize slices
+
+# %% [markdown]
+# With a utility function, [`slice_dataframe`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/slicing/snorkel.slicing.slice_dataframe.html#snorkel.slicing.slice_dataframe), we can visualize examples belonging to this slice in a `pandas.DataFrame`.
+
+# %%
+from snorkel.slicing import slice_dataframe
+
+short_link_df = slice_dataframe(df_valid, short_link)
+short_link_df[["text", "label"]]
+
+# %% [markdown]
+# ## 2. Train a discriminative model
+#
+# To start, we'll initialize a discriminative model using our [`MultitaskClassifier`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html#snorkel.classification.MultitaskClassifier).
+# We'll assume that you are familiar with Snorkel's multitask model — if not, we'd recommend you check out our [Multitask Tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/multitask/multitask_tutorial.ipynb).
```

Comment:
We should not have this dependence.  Ideally the reader has to think about "multi-task" as little as possible; should at least be able to just mention in passing that we use this model, can read about specifically if interested.

**However:** why can't we have a single task, slicing-compatible model so you don't have to deal with any of this MTL if you're just a user that wants to use SFs?  Decent likelihood I'm missing an obvious blocker to this since I've been out of the loop... but if not I feel **very** strongly about this!  @bhancock8